### ISSUE TITLE
Smarter `smart` Extension

### ISSUE THREAD

**Issue by anasram** [jgm/pandoc#6330]
Pandoc version: 2.9.2.1

*****

I'd like to propose the following procedures for a better  `smart` extension:

Case | Original | Normalized
---|---|---
Remove spaces occurs right before punctuation marks. | Hi ! How are you ? | Hi! How are you?
Put the full stop right after the closing quotation mark. | He said: "Yes." | He said: "Yes".<br/>>><br/>He said: “Yes”.
Guess EM & EN dashes from context. | Tom can -he claims- draw 5-10 paintings per day. | Tom can —he claims— draw 5–10 paintings per day.
Localize digits. |&#8235; نسبة النمو 15.5% |&#8235; نسبة النمو ١٥٫٥٪

**Comment by alerque** [comment]
I am a big fan of localization and good typography, but I think Pandoc might be the wrong place in the chain to address these specific issues.

> Remove spaces occurs right before punctuation marks.

This is a language and region specific thing. In French one has to _add_ such spaces to properly typeset punctuation. Normalizing a document to conform with regional typesetting norms is not the role of a format conversion program.

I think the right place to do this would be a filter. You could write it in Lua or Python (or other language) to catch and normalize these things during a Pandoc pass, but the basic function here is not actually document conversion. It provides a nice hook in this case, but it's a a different toolbox.

You _definitely_ don't want to mix this up with the current *smart* extension which handles input format and encoding issues, not content munging.

> Put the full stop right after the closing quotation mark.

I have (somewhere) a pre-processing script that does just this on Markdown content if you are interested, but having run it on a number of large corpses it isn't a straightforward operation and you probably don't want to do this as much as you think you do. Language style guides vary, but most English guides have places where both ways are correct, and determining from context which rule applies is not that easy. Again this is content munging, not document conversion.

> Guess EM & EN dashes from context.

The *smart* extension provides a way to input glyphs many people don't know how to type. What you _don't_ want to generalize is **guessing** what they meant. You will be wrong a lot of the time. I've done this copy editing books and had to had fix hundreds of places I stuffed em/en dashes and it turn out a leading/trailing hyphen was actually correct.


> Localize digits.

You're joking right? This would break _properly_ localized digits more often that fix improper one. What you are suggesting is translation and localization, not document format conversion. I've worked with this one quite a bit dealing with scripts that handle money in localized currencies (see `hledger`) and even with much stricter input formats (such as fixed floating point values) it's still really easy to mess this up. For example is `₺6.200` really six thousand two hundred Turkish Lira (which should probably be entered as `₺6.200,00` for clarity) or is it a typo for `₺6,20`? Don't even get me started on how deep a rabbit hole this is.

Short version: preprocess or filter your own data to your own liking, but these issues do not generalize well and are not a good match for a document format conversion utility.

**Comment by jgm** [comment]
> Put the full stop right after the closing quotation mark.

This is actually wrong in American style.  Quote mark goes after full stop.

Anyway, I agree with @alerque - for people with special interests like this, filters are a good solution.

**Comment by anasram** [comment]
First, thank you @alerque for both detailed and short versions!

Second, I hope my comment here is decent now after closing the issue. Although I almost agree with all of your comments now, I found this discussion very informative and exciting, and I wanted to share more comments after your permission Dr. @jgm 😊

---

> In French one has to add such spaces to properly typeset punctuation.

Likewise in the old school of Arabic typography. Actually I used to thing it became an obsolete way of typing nowadays; as it might cause to show a punctuation mark in a new line, but you're right, such spaces are still used in French.

Yet, notice that in this case, `smart` extension can't convert straight quotes to curly ones:

> Il a dit: " Oui ".

> Normalizing a document to conform with regional typesetting norms is not the role of a format conversion program.

What if such a format conversion program itself generated a mistake in terms of regional typesetting norms? Check issue #6302.

I think it's not easy to separate those two things (format conversion & regional typesetting norms) from each other. 

> You definitely don't want to mix this up with the current smart extension which handles input format and encoding issues, not content munging.

Sure! In fact, I used to use `sed` alongside with Pandoc to handle these issues, but when I started suffering from  some content munging, I thought of submitting this issue here in the hope of _outsourcing_ such a job to Pandoc itself, as Pandoc developers might make it in a smarter way. However, it seems it's not easy to anyone.

> You're joking right?

🤦

Yes! suuuuuuure I was joking 😁

However, please don't tell it to anyone.. it really sounds stupid! 😷

> This is actually wrong in American style. Quote mark goes after full stop.

You're right. Actually, I used to think that web and software industry are driving —over time— to globalize different language styles.

**Comment by alerque** [comment]
> Yet, notice that in this case, `smart` extension can't convert straight quotes to curly ones:
>
> > Il a dit: " Oui ".

This is an interesting case. I would argue it's probably okay to not convert it the exact case you show. There would be too many false positives if you tried. But also note that the _actual_ French style doesn't call for a regular breaking space there! It calls for a non-breaking space, or if I remember correctly a thin non-breaking space. Both of those variants actually work in Pandoc already:

```shell
# Space U+0020
$ pandoc -f markdown+smart -t markdown-smart <<< 'Il a dit: " Oui ".'
Il a dit: " Oui ".

# No-Break Space U+00A0
$ pandoc -f markdown+smart -t markdown-smart <<< 'Il a dit: " Oui ".'
Il a dit: “ Oui ”.

# Narrow No-Break Space U+202F
$ pandoc -f markdown+smart -t markdown-smart <<< 'Il a dit: " Oui ".'
Il a dit: “ Oui ”.

# Thin Space U+2009
$ pandoc -f markdown+smart -t markdown-smart <<< 'Il a dit: " Oui ".'
Il a dit: “ Oui ”.
```

The latter of those might not even be correct as technically it's not a non-breaking space and perhaps shouldn't be handled this way.
